File: rest/src/main/resources/static/src/components/Page.js
import '../styles/style.css';
import React from 'react';
import {Authentication} from './Authentication';
import {CoursesList} from './CoursesList';
import {credentials} from '../scripts';
import {Api} from '../Api';
import Cookies from 'js-cookie';
import ReactDOM from 'react-dom';
import {Notification} from './Notification';

export class Page extends React.Component {

    constructor(props) {
        super(props);

        this.onLogin = this.onLogin.bind(this);
        this.onLogout = this.onLogout.bind(this);

        this.state = {
            account: null
        };

        Api.retrieveAccount(
            credentials(),
            account => {
                this.setState({account: account});
            },
            response => ReactDOM.render(
                <Notification succeed message={"Account retrieving failed.<br/>" + response}/>,
                document.getElementById('notifications')
            )
        );
    }

    render() {
        if (this.state.account != null) {
            return (
                <article className="page">
                    <Authentication account={this.state.account} onLogin={this.onLogin} onLogout={this.onLogout}/>
                    <CoursesList account={this.state.account}/>
                </article>
            );
        } else {
            return (
                <article className="page">
                    <h1>Flaxo educational system</h1>
                    <Authentication account={this.state.account} onLogin={this.onLogin} onLogout={this.onLogout}/>
                </article>
            );
        }
    }

    onLogin(username, password, account) {
        Cookies.set('username', username);
        Cookies.set('password', password);

        this.setState({account: account});
    }

    onLogout() {
        Cookies.remove('username');
        Cookies.remove('password');

        this.setState({account: null});
    }
}

File: frontend/src/main/kotlin/org/flaxo/frontend/component/Page.kt
package org.flaxo.frontend.component

import kotlinx.coroutines.experimental.launch
import react.*
import react.dom.div
import org.flaxo.frontend.Container
import org.flaxo.frontend.client.FlaxoClient
import org.flaxo.frontend.client.FlaxoHttpException
import org.flaxo.frontend.credentials
import org.flaxo.frontend.wrapper.Cookies
import org.flaxo.common.data.User
import org.flaxo.frontend.Notifications

/**
 * Adds application main page.
 */
fun RBuilder.page() = child(Page::class) {}

private class PageState(var user: User? = null) : RState

private class Page : RComponent<EmptyProps, PageState>() {

    private companion object {
        const val USERNAME_COOKIE: String = "username"
        const val PASSWORD_COOKIE: String = "password"
    }

    private val flaxoClient: FlaxoClient

    init {
        flaxoClient = Container.flaxoClient
        launch {
            credentials?.also {
                try {
                    val user = flaxoClient.getSelf(it)
                    setState { this.user = user }
                } catch (e: FlaxoHttpException) {
                    console.log(e)
                    Notifications.error("Error occurred while retrieving ${it.username} user.", e)
                }
            }
        }
    }

    override fun RBuilder.render() {
        val user = state.user
        if (user != null) {
            div("page") {
                courses(user, ::onLogout)
            }
        } else {
            div("page") {
                welcomeDesk(::onLogin)
            }
        }
    }

    private fun onLogin(username: String, password: String, retrievedUser: User) {
        Cookies.set(USERNAME_COOKIE, username)
        Cookies.set(PASSWORD_COOKIE, password)

        setState { user = retrievedUser }
    }

    private fun onLogout() {
        Cookies.remove(USERNAME_COOKIE)
        Cookies.remove(PASSWORD_COOKIE)

        setState { user = null }
    }
}
